Sources/KDTree.swift

```diff
@@ -18,93 +18,121 @@ public enum KDTree<Element: KDTreePoint> {
     indirect case node(left: KDTree<Element>, value: Element, dimension: Int, right: KDTree<Element>)
 }
 
+fileprivate extension UnsafeMutablePointer {
+    @_transparent
+    func swapAt(_ i: Int, _ j: Int) {
+        let temp = self[i]
+        self[i] = self[j]
+        self[j] = temp
+    }
+}
+
 extension KDTree {
     public init(values: [Element], depth: Int = 0) {
         guard !values.isEmpty else {
             self = .leaf
             return
         }
         
-        let currentSplittingDimension = depth % Element.dimensions
-        if values.count == 1, let firstValue = values.first {
-            self = .node(left: .leaf, value: firstValue, dimension: currentSplittingDimension, right: .leaf)
-        }
-        else {
-            var median = values.count / 2
+        let count = values.count
+        
+        let pointer = UnsafeMutablePointer<Element>.allocate(capacity: count)
+        
+        // copy values from the array
+        pointer.initialize(from: values, count: count)
+        
+        self = KDTree(values: pointer, startIndex: 0, endIndex: count)
+        
+        // deallocate the pointer
+        pointer.deallocate()
+    }
+    
+    public init(values: ArraySlice<Element>, depth: Int = 0) {
+
+        let count = values.count
+        
+        self = values.withUnsafeBufferPointer { bufferPointer in
+            guard let start = bufferPointer.baseAddress, count > 0 else {
+                return .leaf
+            }
             
-            let selected = KDTree.quickSelect(targetIndex: median, array: values[0..<values.endIndex], kdDimension: currentSplittingDimension)
-            let medianElement = selected[median]
-            let medianValue = medianElement.kdDimension(currentSplittingDimension)
-          
-            //Ensure left subtree contains currentSplittingDimension-coordinate strictly less than its parent node
-            //Needed for 'contains' and 'removing' method.
-            while median >= 1 && abs(selected[median-1].kdDimension(currentSplittingDimension) - medianValue) < Double.ulpOfOne {
-              median -= 1
+            let pointer = UnsafeMutablePointer<Element>.allocate(capacity: count)
+            
+            defer {
+                // deallocate the pointer later
+                pointer.deallocate()
             }
-          
-            let leftTree = KDTree(values: selected[0..<median], depth: depth+1)
-            let rightTree = KDTree(values: selected[median+1..<selected.count], depth: depth+1)
             
-            self = .node(left: leftTree, value: selected[median],
-                         dimension: currentSplittingDimension, right: rightTree)
+            // copy values from the array
+            pointer.initialize(from: start, count: count)
+            
+            return KDTree(values: pointer, startIndex: 0, endIndex: count)
```

Comment:
this should also pass through the depth parameter
```suggestion
            return KDTree(values: pointer, startIndex: 0, endIndex: count, depth: depth)
```